src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,358 @@
+> {
+  type: 'actionListenerMiddleware/add'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D, O>
+    options?: O
+  }
+}
+
+/**
+ * @alpha
+ */
+export const addListenerAction = createAction(
+  'actionListenerMiddleware/add',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any, any>,
+    options?: ActionListenerOptions<AnyAction, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener,
+        options
+      }
+    }
+  }
+) as BaseActionCreator<
+  {
+    type: string
+    listener: ActionListener<any, any, any, any>
+    options: ActionListenerOptions<any, any, any>
+  },
+  'actionListenerMiddleware/add'
+> & {
+  <
+    C extends TypedActionCreator<any>,
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions<ReturnType<C>, S, D, When>
+  >(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, O>,
+    options?: O
+  ): AddListenerAction<ReturnType<C>, S, D, O>
+
+  <
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions<AnyAction, S, D, When>
+  >(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, O>,
+    options?: O
+  ): AddListenerAction<AnyAction, S, D, O>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D, any>
+  }
+}
+
+/**
+ * @alpha
+ */
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string; listener: ActionListener<any, any, any, any> },
+  'actionListenerMiddleware/remove'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, any>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, any>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+/**
+ * @alpha
+ */
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions<any, S, D, When> & {
+    listener: ActionListener<any, S, D, any>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry> | undefined> = {}
```

Comment:
This might be wishful thinking, but the 1-M action type to listeners relationship almost feels backwards to me.  This becomes clumsy if I wanted to have a listener respond to a set or actions, or all actions especially, whereas the other way round (1-M listener to action types) does not.

Essentially I'm advocating for treating the listener more like a reducer where the action is used to conditionally run logic, and the middleware is just responsible for collecting the set of listeners and passing the actions to them.

There are likely repercussions to changing it around like this for things like the `once` option (how would the middleware know if it fired or not) and is probably more or a rethink of the whole approach than an iteration of this proposal.

Reply:
It sounds like this would be an internal change that on its own wouldn't be exposed to the user.  This map of action type to listeners approach should be more efficient if there are a lot of different action types with listeners attached, but I think you'd need a LOT to make it noticeable.

On the other hand, switching to a list of listeners that each have one or more action types that they are subscribed to would enable pattern-based subscriptions.  For example `addListenerAction('actionListenerMiddleware/*', listener)`.  Heck, if you do that, then you could listen for every action in the world and it would create another way to plug-in something like middleware `addListenerAction('*', listener)`.  I can't decide if I like this.

Reply:
Honestly, at that point you really want to write a middleware by hand, because otherwise we'd provide a middleware that allowed you to write middlewares in the style of writing middlewares.

Reply:
@nogwater So far we've avoided this string-pattern matching everywhere else in RTK, and for good reason. Once you fall back to that, TypeScript really loses all type safety as every action just boils down to `any`.

Reply:
@phryneas, yeah as I said, that idea is probably a competing proposal more than an iteration of this one.

In my ideal implementation, the listener middleware would be a more reduxy implementation that leaned into the more functional concepts and loose action/reducer coupling while RTK would provide the opinionated, type-safe utilities to make writing listeners for the 95% use-cases straight forward and easier (in a similar way to how `redux-thunk` is supported).